src/core/search/block_list.cc

```diff
@@ -0,0 +1,128 @@
+#include "core/search/block_list.h"
+
+namespace dfly::search {
+
+using namespace std;
+
+template <typename C> bool BlockList<C>::Insert(DocId t) {
+  auto block = FindBlock(t);
+  if (block == blocks_.end())
+    block = blocks_.insert(blocks_.end(), C{blocks_.get_allocator().resource()});
+
+  if (!block->Insert(t))
+    return false;
+
+  size_++;
+  TrySplit(block);
+  return true;
+}
+
+template <typename C> bool BlockList<C>::Remove(DocId t) {
+  if (auto block = FindBlock(t); block != blocks_.end() && block->Remove(t)) {
+    size_--;
+    TryMerge(block);
+    return true;
+  }
+
+  return false;
+}
```

Comment:
The other option is not compacting blocks and using a separate GC phase instead (like redis does). This also has the benefit that blocks don't have to be split on 2*K, but can be split directly on K+1 (as they won't be compacted immediately)... But I'm not sure about that 🤔